## 001_KOSCOM data 날짜 - 시간 붙여서 새 컬럼 만들기

In [1]:
install.packages("dplyr")
library("dplyr")

package 'dplyr' successfully unpacked and MD5 sums checked


Warning message:
"cannot remove prior installation of package 'dplyr'"


The downloaded binary packages are in
	C:\Temp\RtmpEBGuEK\downloaded_packages



Attaching package: 'dplyr'

The following objects are masked from 'package:stats':

    filter, lag

The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union



In [20]:
cust <- read.csv('C:\\Users\\진수빈\\Desktop\\customer_0813_0817_TOP500_no_quotation_v2\\CUSTOMER_ISSUE_V2.csv')

In [3]:
head(cust)

CUSTOMER_ID ISSUE_CD ISSUE_NAME MARKET_ID QUERY_DATE QUERY_TIME
1 CUST_1      005930   삼성전자   1         20180813   70124     
2 CUST_1      005930   삼성전자   1         20180816   62415     
3 CUST_1      005930   삼성전자   1         20180816   62416     
4 CUST_1      005930   삼성전자   1         20180816   62420     
5 CUST_1      005930   삼성전자   1         20180816   62426     
6 CUST_1      005930   삼성전자   1         20180817   70348

In [5]:
table(nchar(as.character(cust$QUERY_DATE)))


      0       2       8       9 
   1560    1252 1308608      24 

In [6]:
table(nchar(cust$QUERY_TIME))


     2      4      5      6      8 
    24     82 362602 945924   1252 

## 1. 데이터 문제 발생
### 1. QUERY_DATE의 길이는 8이어야 하는데 0, 2 9 존재
### 2. QUERY_TIME의 길이는 6이어야 하는데 2, 4, 5, 8 존재
# 
## 2. 문제 원인 탐색
### 1. QT = 8, QD = 2가 1252개로 같음 -> 데이터가 밀렸을 것으로 추정
### 2. QD=9, QT = 2가 24개로 같음 -> 마찬가지로 밀렸을 것임
#### * 1, 2 케이스는 똑같이 밀렸지만 밀린 칸 수가 다를 것으로 추정
#
### 3. QD = 0 인 경우는 무슨 문제인지 추정 불가능
### 4. QD = 8의 수와 (QT = 4) + (QT = 5) + (QT = 6)이 같음
#### * QT = 4인 경우 00시인 경우로 추정
#### * QT = 4인 경우 0x시인 경우로 추정

## 3. 문제 해결
### 1. QD = 0 인 경우

In [11]:
head(filter(cust, nchar(as.character(cust$QUERY_DATE)) == 0))
tail(filter(cust, nchar(as.character(cust$QUERY_DATE)) == 0))

CUSTOMER_ID ISSUE_CD ISSUE_NAME  
1 CUST_118    SMB@EUR  유로통화    
2             112      20180814    
3 CUST_118    SMB@EUR  유로통화    
4             112      20180814    
5 CUST_118    SMB@JPY  일본_엔(100)
6             112      20180817    
  MARKET_ID                                                                                                                                                      
1                                                                                                                                                                
2 80432                                                                                                                                                          
3                                                                                                                                                                
4 80433                                                                                                                                                          
5                                                                                                                                                                
6 73855                                                                                                                                                          
  QUERY_DATE QUERY_TIME
1            NA        
2            NA        
3            NA        
4            NA        
5            NA        
6            NA

CUSTOMER_ID                                                                                                                           
1555 75700                                                                                                                                 
1556 104420                                                                                                                                
1557 75841                                                                                                                                 
1558 74827                                                                                                                                 
1559 91118                                                                                                                                 
1560 100738                                                                                                                                
     ISSUE_CD ISSUE_NAME MARKET_ID QUERY_DATE QUERY_TIME
1555                                          NA        
1556                                          NA        
1557                                          NA        
1558                                          NA        
1559                                          NA        
1560                                          NA

### head()의 경우의 데이터가 임의의 개행으로 밀린 것을 확인 할 수 있음
### tail()의 경우 시간 값으로 추정되고 밀린 데이터가 개행 된 것으로 추정
### head의 경우 엑셀로 작업 wrong.csv로 전처리
### taild의 경우 데이터가 2칸 밀린 경우와 1칸 밀린 경우로 나누어  처리
#### * 2칸 밀린 경우 date_time.csv로 처리
#### * 1칸 밀린 경우 time.csv로 처리
#### * 밀린 애들은 나중에 당길때 뒤에 붙일 예정

In [63]:
wrong <- read.csv('C:\\Users\\진수빈\\Desktop\\customer_0813_0817_TOP500_no_quotation_v2\\wrong.csv')
head(wrong)

X CUSTOMER_ID ISSUE_CD ISSUE_NAME   MARKET_ID QUERY_DATE QUERY_TIME
1 1 CUST_118    SMB@EUR  유로통화     112       20180814   80432     
2 2 CUST_118    SMB@EUR  유로통화     112       20180814   80433     
3 3 CUST_118    SMB@JPY  일본_엔(100) 112       20180817   73855     
4 4 CUST_118    SMB@JPY  일본_엔(100) 112       20180814   80432     
5 5 CUST_118    SMB@JPY  일본_엔(100) 112       20180813   90331     
6 6 CUST_118    SMB@JPY  일본_엔(100) 112       20180814   75549

In [16]:
time_date <- read.csv('C:\\Users\\진수빈\\Desktop\\customer_0813_0817_TOP500_no_quotation_v2\\time_date.csv')
head(time_date)

癤풯UERY_DATE QUERY_TIME
1 20180814      100738    
2 20180816      162110    
3 20180813       74826    
4 20180816       94243    
5 20180813       91117    
6 20180817      104419

In [46]:
time2 <- read.csv('C:\\Users\\진수빈\\Desktop\\customer_0813_0817_TOP500_no_quotation_v2\\time.csv')
head(time)

癤풯UERY_TIME
1 111101       
2 172431       
3 111201       
4 111158       
5 172431       
6 111101

### 2. QD = 9인 경우
#### MI, QD 자리에 ISSUE_NAME이 들어간 것 확인
#### 2칸 밀린 경우임 -> date_time.csv와 병합 가능

In [18]:
head(filter(cust, nchar(as.character(cust$QUERY_DATE)) == 9))

CUSTOMER_ID ISSUE_CD ISSUE_NAME            MARKET_ID QUERY_DATE QUERY_TIME
1 CUST_60     211N4037 삼성전자   C 201804 2 450       000(  10)  12        
2 CUST_60     211N4037 삼성전자   C 201804 2 450       000(  10)  12        
3 CUST_60     211N4037 삼성전자   C 201804 2 450       000(  10)  12        
4 CUST_60     211N4037 삼성전자   C 201804 2 450       000(  10)  12        
5 CUST_60     211N4037 삼성전자   C 201804 2 450       000(  10)  12        
6 CUST_60     211N4037 삼성전자   C 201804 2 450       000(  10)  12

#### 두칸 밀린 데이터를 cust2로 따로 저장하고 처리

In [21]:
cust2 <- filter(cust, nchar(as.character(cust$QUERY_DATE)) == 9)
cust2$ISSUE_NAME <- paste0(cust2$ISSUE_NAME, cust2$MARKET_ID, cust2$QUERY_DATE)
head(cust2)

CUSTOMER_ID ISSUE_CD ISSUE_NAME                        MARKET_ID QUERY_DATE
1 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 450       000(  10) 
2 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 450       000(  10) 
3 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 450       000(  10) 
4 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 450       000(  10) 
5 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 450       000(  10) 
6 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 450       000(  10) 
  QUERY_TIME
1 12        
2 12        
3 12        
4 12        
5 12        
6 12

In [24]:
cust2$MARKET_ID <- cust2$QUERY_TIME
head(cust2)

CUSTOMER_ID ISSUE_CD ISSUE_NAME                        MARKET_ID QUERY_DATE
1 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        000(  10) 
2 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        000(  10) 
3 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        000(  10) 
4 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        000(  10) 
5 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        000(  10) 
6 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        000(  10) 
  QUERY_TIME
1 12        
2 12        
3 12        
4 12        
5 12        
6 12

In [27]:
cust2$QUERY_DATE <- time_date$癤풯UERY_DATE
cust2$QUERY_TIME <- time_date$QUERY_TIME

In [28]:
head(cust2)

CUSTOMER_ID ISSUE_CD ISSUE_NAME                        MARKET_ID QUERY_DATE
1 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        20180814  
2 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        20180816  
3 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        20180813  
4 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        20180816  
5 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        20180813  
6 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        20180817  
  QUERY_TIME
1 100738    
2 162110    
3  74826    
4  94243    
5  91117    
6 104419

#### QD = 9 인 경우 처리 완료

### 3. QD = 2 인 경우

In [43]:
cust1 <- filter(cust, nchar(as.character(cust$QUERY_DATE)) == 2)
head(cust1)

CUSTOMER_ID ISSUE_CD ISSUE_NAME                MARKET_ID QUERY_DATE
1 CUST_163    211N9033 삼성전자   C 201809    45 000(  10) 12        
2 CUST_163    211N9033 삼성전자   C 201809    45 000(  10) 12        
3 CUST_163    211N9033 삼성전자   C 201809    45 000(  10) 12        
4 CUST_163    2C6N9021 아모레퍼시 C 201809   260 000(  10) 12        
5 CUST_163    2C6N9021 아모레퍼시 C 201809   260 000(  10) 12        
6 CUST_163    2C6N9021 아모레퍼시 C 201809   260 000(  10) 12        
  QUERY_TIME
1 20180816  
2 20180814  
3 20180816  
4 20180816  
5 20180814  
6 20180816

In [44]:
cust1$ISSUE_NAME <- paste0(cust1$ISSUE_NAME, cust1$MARKET_ID)
cust1$MARKET_ID <- cust1$QUERY_DATE
cust1$QUERY_DATE <- cust1$QUERY_TIME
head(cust1)

CUSTOMER_ID ISSUE_CD ISSUE_NAME                         MARKET_ID QUERY_DATE
1 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180816  
2 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180814  
3 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180816  
4 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180816  
5 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180814  
6 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180816  
  QUERY_TIME
1 20180816  
2 20180814  
3 20180816  
4 20180816  
5 20180814  
6 20180816

In [47]:
cust1$QUERY_TIME <- time2$癤풯UERY_TIME
head(cust1)

CUSTOMER_ID ISSUE_CD ISSUE_NAME                         MARKET_ID QUERY_DATE
1 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180816  
2 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180814  
3 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180816  
4 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180816  
5 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180814  
6 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180816  
  QUERY_TIME
1 111101    
2 172431    
3 111201    
4 111158    
5 172431    
6 111101

#### QD = 2 처리 완료

## 4. DATA 결합 & 확인

In [53]:
cust3 <- filter(cust, nchar(as.character(cust$QUERY_DATE)) == 8)
head(cust3)

CUSTOMER_ID ISSUE_CD ISSUE_NAME MARKET_ID QUERY_DATE QUERY_TIME
1 CUST_1      005930   삼성전자   1         20180813   70124     
2 CUST_1      005930   삼성전자   1         20180816   62415     
3 CUST_1      005930   삼성전자   1         20180816   62416     
4 CUST_1      005930   삼성전자   1         20180816   62420     
5 CUST_1      005930   삼성전자   1         20180816   62426     
6 CUST_1      005930   삼성전자   1         20180817   70348

In [55]:
head(cust1)
head(cust2)
head(cust3)
head(wrong)

CUSTOMER_ID ISSUE_CD ISSUE_NAME                         MARKET_ID QUERY_DATE
1 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180816  
2 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180814  
3 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180816  
4 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180816  
5 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180814  
6 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180816  
  QUERY_TIME
1 111101    
2 172431    
3 111201    
4 111158    
5 172431    
6 111101

CUSTOMER_ID ISSUE_CD ISSUE_NAME                        MARKET_ID QUERY_DATE
1 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        20180814  
2 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        20180816  
3 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        20180813  
4 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        20180816  
5 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        20180813  
6 CUST_60     211N4037 삼성전자   C 201804 2450000(  10) 12        20180817  
  QUERY_TIME
1 100738    
2 162110    
3  74826    
4  94243    
5  91117    
6 104419

CUSTOMER_ID ISSUE_CD ISSUE_NAME MARKET_ID QUERY_DATE QUERY_TIME
1 CUST_1      005930   삼성전자   1         20180813   70124     
2 CUST_1      005930   삼성전자   1         20180816   62415     
3 CUST_1      005930   삼성전자   1         20180816   62416     
4 CUST_1      005930   삼성전자   1         20180816   62420     
5 CUST_1      005930   삼성전자   1         20180816   62426     
6 CUST_1      005930   삼성전자   1         20180817   70348

X CUSTOMER_ID ISSUE_CD ISSUE_NAME   MARKET_ID QUERY_DATE QUERY_TIME
1 1 CUST_118    SMB@EUR  유로통화     112       20180814   80432     
2 2 CUST_118    SMB@EUR  유로통화     112       20180814   80433     
3 3 CUST_118    SMB@JPY  일본_엔(100) 112       20180817   73855     
4 4 CUST_118    SMB@JPY  일본_엔(100) 112       20180814   80432     
5 5 CUST_118    SMB@JPY  일본_엔(100) 112       20180813   90331     
6 6 CUST_118    SMB@JPY  일본_엔(100) 112       20180814   75549

In [64]:
wrong <- subset(wrong, select = c(-X))

In [65]:
head(wrong)

CUSTOMER_ID ISSUE_CD ISSUE_NAME   MARKET_ID QUERY_DATE QUERY_TIME
1 CUST_118    SMB@EUR  유로통화     112       20180814   80432     
2 CUST_118    SMB@EUR  유로통화     112       20180814   80433     
3 CUST_118    SMB@JPY  일본_엔(100) 112       20180817   73855     
4 CUST_118    SMB@JPY  일본_엔(100) 112       20180814   80432     
5 CUST_118    SMB@JPY  일본_엔(100) 112       20180813   90331     
6 CUST_118    SMB@JPY  일본_엔(100) 112       20180814   75549

In [91]:
newcust <- rbind(cust1, cust2, cust3, wrong)

Warning message in `[<-.factor`(`*tmp*`, ri, value = c(112L, 112L, 112L, 112L, 112L, :
"invalid factor level, NA generated"

In [92]:
table(nchar(as.character(newcust$QUERY_DATE) == 8))


      5 
1310026 

In [93]:
table(nchar(as.character(newcust$QUERY_TIME)))


     4      5      6 
    82 363453 946491 

#### QT = 4인 경우 00 붙임
#### QT = 5인 경우 0 붙임

In [94]:
newcust$QUERY_TIME <- ifelse(nchar(as.character(newcust$QUERY_TIME)) == 4, paste0("00", as.character(newcust$QUERY_TIME)), newcust$QUERY_TIME)

In [95]:
newcust$QUERY_TIME <- ifelse(nchar(as.character(newcust$QUERY_TIME)) == 5, paste0("0", as.character(newcust$QUERY_TIME)), newcust$QUERY_TIME)

In [96]:
table(nchar(as.character(newcust$QUERY_TIME)))


      6 
1310026 

#### 아무 문제 없음

## 5. 날짜 - 시간 형식 변환 

In [97]:
newcust <- transform(newcust, X2 = paste(substr(newcust$QUERY_TIME, 1, 2), substr(newcust$QUERY_TIME, 3, 4), substr(newcust$QUERY_TIME, 5, 6), sep = ":"))
head(newcust)

CUSTOMER_ID ISSUE_CD ISSUE_NAME                         MARKET_ID QUERY_DATE
1 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180816  
2 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180814  
3 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180816  
4 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180816  
5 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180814  
6 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180816  
  QUERY_TIME X2      
1 111101     11:11:01
2 172431     17:24:31
3 111201     11:12:01
4 111158     11:11:58
5 172431     17:24:31
6 111101     11:11:01

In [98]:
newcust <- transform(newcust, Y2 = paste(substr(newcust$QUERY_DATE, 1, 4), substr(newcust$QUERY_DATE, 5, 6), substr(newcust$QUERY_DATE, 7, 8), sep = "-"))
head(newcust)

CUSTOMER_ID ISSUE_CD ISSUE_NAME                         MARKET_ID QUERY_DATE
1 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180816  
2 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180814  
3 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180816  
4 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180816  
5 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180814  
6 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180816  
  QUERY_TIME X2       Y2        
1 111101     11:11:01 2018-08-16
2 172431     17:24:31 2018-08-14
3 111201     11:12:01 2018-08-16
4 111158     11:11:58 2018-08-16
5 172431     17:24:31 2018-08-14
6 111101     11:11:01 2018-08-16

In [99]:
newcust$DATE_TIME <- paste(newcust$Y2, newcust$X2)
head(newcust)

CUSTOMER_ID ISSUE_CD ISSUE_NAME                         MARKET_ID QUERY_DATE
1 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180816  
2 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180814  
3 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180816  
4 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180816  
5 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180814  
6 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180816  
  QUERY_TIME X2       Y2         DATE_TIME          
1 111101     11:11:01 2018-08-16 2018-08-16 11:11:01
2 172431     17:24:31 2018-08-14 2018-08-14 17:24:31
3 111201     11:12:01 2018-08-16 2018-08-16 11:12:01
4 111158     11:11:58 2018-08-16 2018-08-16 11:11:58
5 172431     17:24:31 2018-08-14 2018-08-14 17:24:31
6 111101     11:11:01 2018-08-16 2018-08-16 11:11:01

In [100]:
newcust <- subset(newcust, select = c(-X2, -Y2))
head(newcust)

CUSTOMER_ID ISSUE_CD ISSUE_NAME                         MARKET_ID QUERY_DATE
1 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180816  
2 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180814  
3 CUST_163    211N9033 삼성전자   C 201809    45000(  10) 12        20180816  
4 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180816  
5 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180814  
6 CUST_163    2C6N9021 아모레퍼시 C 201809   260000(  10) 12        20180816  
  QUERY_TIME DATE_TIME          
1 111101     2018-08-16 11:11:01
2 172431     2018-08-14 17:24:31
3 111201     2018-08-16 11:12:01
4 111158     2018-08-16 11:11:58
5 172431     2018-08-14 17:24:31
6 111101     2018-08-16 11:11:01

#### 끝